homeassistant/components/osoenergy/water_heater.py

```diff
@@ -0,0 +1,194 @@
+"""Support for OSO Energy water heaters."""
+from collections.abc import Mapping
+from typing import Any
+
+from homeassistant.components.water_heater import (
+    WaterHeaterEntity,
+    WaterHeaterEntityFeature,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import STATE_OFF, STATE_ON, UnitOfTemperature
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.device_registry import DeviceInfo
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+import homeassistant.util.dt as dt_util
+
+from . import OSOEnergyEntity
+from .const import DOMAIN
+
+EXTRA_HEATER_ATTR: dict[str, dict[str, Any]] = {
```

Comment:
I discussed this with @MartinHjelmare 

The general guideline for state attributes is:
- If the measurement makes sense without knowing the main state of the water heater is should be a separate entity.
- If you have to know the main state to make sense of the measurement it should be a state attribute.

Based on this, please make sensors for all except the heater state which should be binary sensor. The mode and optimization mode should be enum sensors.

Reply:
I have removed the extra state attributes. It is my understanding that the scope of this pull request should be as small as possible and sensors and scripts should be defined in another pull request.